Agrega codigo necesario por mongoDB setup y spawn en Windows (#1353)
Laboratoria/bootcamp: projects/04-burger-queen-api/guides/GETTING-STARTED-MONGODB.md

Estes servicios tienen que estar corriendo, disponible, para este proyecto,
el API, puede funcionar.

Para correr el server, es bien simple - con `npm start` va a ejecutar
[`index.js`](../index.js) que tiene logica basica de server con express.

Para levantar el server, este depende en tu instalación y sistema.
Por ejemplo, en MacOS si instalaste con `homebrew`, puede usar
`brew services start mongodb-community@6.0` para levantarlo. O sin `homebrew`,
`mongod --config /usr/local/etc/mongod.conf --fork`. En Windows, hay que
levantarlo desde Services console.

Revisa [la guia de instalacion de
](https://www.mongodb.com/docs/manual/tutorial/install-mongodb-on-os-x/#run-mongodb-community-edition),
[guia de instalacion de Windows](https://www.mongodb.com/docs/manual/tutorial/install-mongodb-on-windows/#run-mongodb-community-edition-as-a-windows-service),
o tu instalación en particular, para ejemplos en como levantarlo.

## 3. Elegir módulo (cliente)

Ahora que ya tenemos un servidor de bases de datos vamos a necesitar elegir un
módulo o librería diseñado para interactuar con nuestra base de datos desde
Node.js. Existen un montón de opciones, pero para este proyecto te recomendamos

> elegir [Mongoose](https://mongoosejs.com/) (que es la más popular para MongoDB).

Author: No tengo opinion de eso...

> elegir [Mongoose](https://mongoosejs.com/) (que es la más popular para MongoDB).

Author: Quite mongoose en favor de node driver. Hable con @lupomontero de eso.

el mundo real, para ello la aplicación de prueba necesitará una base de datos y
escuchar en un puerto de red.

Para correr pruebas e2e sobre instancia local podemos usar:

```sh
npm run test:e2e
```

Esto levanta la aplicación con `npm start` y corre los tests contra la URL de
esta instancia (por defecto `http://127.0.0.1:8080`). Esto asume que la base de
datos está disponible.

Los pruebas e2e depende en codigo de (`globalSetup.js`)[../e2e/globalSetup.js]
que `jest` sabe correr primero antes de los tests.
En este paso de setup, hay que levantar un mock db (como hemos hablado de
[`mongodb-memory-server`](https://github.com/nodkz/mongodb-memory-server))
y hacer que jest se conecta a este mock db.

Para este configuracion - hay que instalar `mongodb-memory-server` y un preset (`jest-mongodb`)[https://github.com/shelfio/jest-mongodb].

El patrón de setup no es tan directo y documentado, porque `jest-mongodb`
tambien usa su propio `globalSetup` y los dos no juegan bien en conjunto.
Hemos agregado este codigo para configurar los e2e tests y `globalSetup`
en un patch (`usando git diff`) que puedes aplicar automaticamente con

> `git apply ./mongoDB-e2e.patch`.

Reviewer: me encanto la opcion de los patch, pero si mongo va a ser la unica opcion por ahora, q el boilerplate venga con el patch ya aplicado

Author: Entonces estamos dicidiendo para ellas que es hecho en Mongo y no con docker localmente. Creo si usan docker van a hacer un poco diferente el `globalSetup` - pero hay que verificar.

# Primeros pasos con MongoDB

> MongoDB es un base de datos no-relacionado, y es la más _común_ (popular)

Reviewer: ```suggestion
MongoDB es un base de datos no-relacional, y es la más _común_ (popular)
```

a día de hoy en el ecosistema de Node.js.

> Si vas a trabajar con MongoDB y Docker, mejor leer

Reviewer: ```suggestion
Si vas a trabajar con MongoDB y Docker, mejor lee primero
```

> [**la guía de _primeros pasos con Docker_**](./GETTING-STARTED-DOCKER.md)

Reviewer: ```suggestion
[**la guía de _primeros pasos con Docker_**](./GETTING-STARTED-DOCKER.md).
```

> primero. Si vas a instalar MongoDB y desarollar el proyecto sin

Reviewer: ```suggestion
Si vas a instalar MongoDB y desarollar el proyecto sin
```

> Docker en este momento, seguir leyendo este guia.

Reviewer: ```suggestion
Docker en este momento, sigue leyendo esta guía.
```

> Una vez creado tu fork y clonado el repo en tu computadora, y haces `npm install`

Reviewer: ```suggestion
Una vez creado tu fork y clonado el repo en tu computadora, haces `npm install`
```

> para instalar las dependencias basicas, antes de poder

Reviewer: ```suggestion
para instalar las dependencias básicas.
```

> comenzar a codear, tenemos que crear nuestro _entorno de desarrollo_.

Reviewer: ```suggestion
Antes de poder comenzar a codear, tenemos que crear nuestro _entorno de desarrollo_.
```

> Para ello te recomendamos seguir los pasos a continuación:

Reviewer: ```suggestion
Para ello te recomendamos seguir los siguientes pasos:
```

* [1. Instalar MongoDB y Compass](#1-instalar-mongodb-y-compass)

> * [2. Levanta el base de datos y server](#2-levanta-el-base-de-datos-y-server)

Reviewer: ```suggestion
* [2. Levanta la base de datos y servidor](#2-levanta-la-base-de-datos-y-servidor)
```

* [3. Elegir módulo (cliente)](#3-elegir-módulo-cliente)
* [4. Definir esquemas](#4-definir-esquemas)
* [5. Implementar los primeros TODOs](#5-implementar-los-primeros-todos)
* [6. Definir estrategia de pruebas unitarias](#6-definir-estrategia-de-pruebas-unitarias)
* [7. Familiarizarte con las pruebas de integración (e2e)](#7-familiarizarte-con-las-pruebas-de-integración-e2e)

***

## 1. Instalar MongoDB y Compass

> El sitio de MongoDB tiene tutorials como instalar el [Community Edition](https://www.mongodb.com/docs/compass/current/install/).

Reviewer: ```suggestion
El sitio de MongoDB tiene tutoriales de cómo instalar la [Community Edition](https://www.mongodb.com/docs/compass/current/install/).
```

Elige tu plataforma (Windows, Mac, Ubuntu etc) y sigue [el tutorial](https://www.mongodb.com/docs/manual/installation/#mongodb-installation-tutorials).

> Con los tutorials, va a descargar y instalar MongoDB en tu sistema,

Reviewer: ```suggestion
Con los tutoriales, vas a descargar y instalar MongoDB en tu sistema,
```

> con algunas herramientas y comandos para usar en la linea de comando.

Reviewer: ```suggestion
con algunas herramientas y comandos para usar en la línea de comando.
```

Tambien recomendamos que instales [Compass](https://www.mongodb.com/products/compass),

> que es un GUI (Graphical User Interface) para interactuar con el base de datos.

Reviewer: ```suggestion
que es un GUI (Graphical User Interface) para interactuar con la base de datos.
```

Puedes interactuar con MongoDB sin Compass y solo en la linea de comando,

> pero un GUI puede ayudarte visualizar y entender que esta en el db. Sigue [las

Reviewer: ```suggestion
pero un GUI puede ayudarte visualizar y entender lo que está en la base de datos. Sigue [las
```

instrucciones de instalacion de Compass](https://www.mongodb.com/docs/compass/current/install/).

> ## 2. Levanta el base de datos y server

Reviewer: ```suggestion
## 2. Levanta la base de datos y server
```

> Podemos decir este proyecto tiene dos "servicios", uno es el base de datos para

Reviewer: ```suggestion
Podemos decir que este proyecto tiene dos "servicios", uno es la base de datos para
```

> almanecenar los productos, usuarios, etc., y el otro es el server para exponer

Reviewer: ```suggestion
almanecenar los productos, usuarios, etc., y el otro es el servidor para exponer
```

el API.

> Estes servicios tienen que estar corriendo, disponible, para este proyecto,

Reviewer: ```suggestion
Estos servicios tienen que estar corriendo, para que
```

> el API, puede funcionar.

Reviewer: ```suggestion
el API pueda funcionar.
```

> Para correr el server, es bien simple - con `npm start` va a ejecutar

Reviewer: ```suggestion
Correr el servidor es bien simple: con `npm start` va a ejecutar
```

> [`index.js`](../index.js) que tiene logica basica de server con express.

Reviewer: ```suggestion
[`index.js`](../index.js) que tiene la lógica básica de un servidor con express.
```

> Para levantar el server, este depende en tu instalación y sistema.

Reviewer: ```suggestion
Cómo levantar el servidor, este depende de tu instalación y sistema.
```

Por ejemplo, en MacOS si instalaste con `homebrew`, puede usar
`brew services start mongodb-community@6.0` para levantarlo. O sin `homebrew`,
`mongod --config /usr/local/etc/mongod.conf --fork`. En Windows, hay que
levantarlo desde Services console.

Revisa [la guia de instalacion de
](https://www.mongodb.com/docs/manual/tutorial/install-mongodb-on-os-x/#run-mongodb-community-edition),
[guia de instalacion de Windows](https://www.mongodb.com/docs/manual/tutorial/install-mongodb-on-windows/#run-mongodb-community-edition-as-a-windows-service),

> o tu instalación en particular, para ejemplos en como levantarlo.

Reviewer: ```suggestion
o tu instalación en particular, para ejemplos en cómo levantarlo.
```

## 3. Elegir módulo (cliente)

Ahora que ya tenemos un servidor de bases de datos vamos a necesitar elegir un
módulo o librería diseñado para interactuar con nuestra base de datos desde
Node.js. Existen un montón de opciones, pero para este proyecto te recomendamos
usar el [Node driver de MongoDB](https://www.mongodb.com/docs/drivers/node/current/)

>  (lo mas directo interactuar con MongoDB).

Reviewer: ```suggestion
que es la forma más directa de interactuar con tu instancia de MongoDB.
```

Hay que instalar el Node Driver en este proyecto usando `npm`.

El _boilerplate_ ya incluye un archivo `config.js` donde se leen las
variables de entorno, y entre ellas está `DB_URL`. Como vemos ese valor lo
estamos asignando en la propiedad `dbUrl` del módulo `config`.

```js
// `config.js`
exports.dbUrl = process.env.DB_URL || "mongodb://localhost:27017/test";
```

Ahora que ya sabemos dónde encontrar el _connection string_ (en el módulo
config), podemos proceder a establecer una conexión con la base de datos
usando el cliente elegido.

### Ejemplo de conexión usando MongoDB Node Driver

```js
const { MongoClient } = require('mongodb');
const config = require("./config");

const client = new MongoClient(config.dbUrl);

async function run() {
  try {
    await client.connect();
    const db = client.db('test');
    // ...
  } finally {
    // Close the database connection when finished or an error occurs
    await client.close();
  }
}
run().catch(console.error);
```

> Puedes encontrar mas ejemplos en [la documentación de MongoDB](https://www.mongodb.com/docs/drivers/node/current/?utm_campaign=w3schools_mdb&utm_source=w3schools&utm_medium=referral#fundamentals)

Reviewer: ```suggestion
Puedes encontrar mas ejemplos en [la documentación de MongoDB](https://www.mongodb.com/docs/drivers/node/current/)
```

> y también en [un tutorial de w3schools](https://www.w3schools.com/mongodb/mongodb_nodejs_connect_database.php).

Author: Jaja ok. Si soy, pero puedo quitarlo.

## 4. Definir esquemas

Como parte del proceso de diseño de nuestra base de datos vamos a tener que
especificar los _esquemas_ de nuestros _modelos_ de datos. Con esto nos
referimos a que tenemos que _describir_ de alguna forma las colecciones o

> tablas que vamos a usar y la _forma_ de los objetos o filas que vayamos a

Reviewer: ```suggestion
que vamos a usar y la _forma_ de los objetos que vayamos a
```

guardar en esas colecciones.

Puedes usar [la documentación de API](https://app.swaggerhub.com/apis-docs/ssinuco/BurgerQueenAPI/2.0.0)
que describe la estructura de `products`, `orders`, etc. para guiar el diseño.

## 5. Implementar primeros TODOs

El boilerplate del proyecto viene con archivos con logica para arrancar el server
y otros que contienen funciones de rutas y authenticación, pero muchas estan vacios.

> Hemos marcado los primeros partes essencial con comentarios `TODO` - que es un

Reviewer: ```suggestion
Hemos marcado las primeras partes esenciales con comentarios `TODO` (del inglés _por hacer_), que es una
```

convención de documentacion de desarollo cuando hay algo para hacer.

> Aqui te guiamos un poco de estes todos.

Reviewer: ```suggestion
Aquí te guiamos un poco sobre esto ToDo's.
```

> ### TODO 1: Conectar al base de datos

Reviewer: ```suggestion
### TODO 1: Conectar a la base de datos
```

En el `index.js` donde arrancamos express y el API, hay

```js
const { port, dbUrl, secret } = config;
const app = express();

// TODO: Conexión a la Base de Datos (MongoDB o MySQL)
```

Aquí debes usar el `dbUrl` que importamos del config para establecer la conexión,
quiza atravez una funcion que defines en un modulo y importas en el index.

### TODO 2: Crear el usuario admin

EL proyecto depende en la existencia de un usuario en el base de datos que
tiene privilegios de administrador - para crear otros usuarios, etc.

En `routes/users.js` invocamos una funcion `initAdminUser(app, next)`
al final de archivo, y definimos esta funcion arriba en este misma archivo.

`initAdminUser` es incompleto, y hay un TODO para crear el usuario admin,

> donde tienes que primero porbar si un admin ya existe, y si no agregar uno

Reviewer: ```suggestion
donde tienes que primero chequear si un admin ya existe, y si no agregar uno
```

> al base de datos.

Reviewer: ```suggestion
a la base de datos.
```

```js
const initAdminUser = (app, next) => {
  const { adminEmail, adminPassword } = app.get('config');
  if (!adminEmail || !adminPassword) {
    return next();
  }

  const adminUser = {
    email: adminEmail,
    password: bcrypt.hashSync(adminPassword, 10),
    roles: { admin: true },
  };

  // TODO: crear usuaria admin
  next();
};

> Puedes confirmar si tu codigo funciona revisando el base de datos y un testeo unitario.

Reviewer: ```suggestion
Puedes confirmar si tu código funciona revisando la base de datos y con un testeo unitario.
```

```

### TODO 3: Implementar autenticación de usuario

En `routes/auth.js` hay la ruta '/auth' definida, con un

```js
 // TODO: autenticar a la usuarix
 ```

> Aqui es donde debe que verificar que el correo y password coinciden con

Reviewer: ```suggestion
Aqui es donde debes verificar que el correo y password coinciden con
```

> datos de un usuario en el base de datos, y si coinciden, hay que generar un

Reviewer: ```suggestion
datos de algún usuario en la base de datos. Si coinciden, hay que generar un
```

[JWT token](https://jwt.io/introduction)

> y mandarlo en la respuesta, para el usuario puede usar con peticiones futuros.

Reviewer: ```suggestion
y mandarlo en la respuesta, para que el usuario puede usarlo en sus peticiones futuras.
```

Para ejemplos con mas detalle, busca tutoriales de autenticacion con JWT y node/express.

### TODO 4: Implementar el middleware de autenticación

> En `middleware/auth.js` hay varios `TODOs`. Usa este oportunidad a familiarizarte

Reviewer: ```suggestion
En `middleware/auth.js` hay varios `TODOs`. Usa esta oportunidad para familiarizarte
```

con [el concepto de middleware en express](https://expressjs.com/es/guide/using-middleware.html).

> La aplicacion va a usar este middleware para chequear que los peticiones

Reviewer: ```suggestion
La aplicación va a usar este middleware para chequear que las peticiones
```

> que requiere autenticacion son autorizados, que tiene un token valido.

Reviewer: ```suggestion
que requieren autenticación sean autorizados, es decir, que tienen un token válido.
```

### TODO 5: Implemenatar la ruta GET `/users`

Para juntar y verificar todo el trabajo que has hecho, seria bueno implementar

> una ruta basica del API, en este caso recomendamos `/users` porque ya debes

Reviewer: ```suggestion
una ruta básica del API, en este caso recomendamos `/users` porque ya debes
```

> tener el user admin en tu base de datos, y este ruta usa el middleware de auth.

Reviewer: ```suggestion
tener el user admin en tu base de datos, y además porque esta ruta usa el middleware de auth.
```

> Vas a ver que la ruta `/users` usa la funcion `getUsers` que es definido en

Reviewer: ```suggestion
Vas a ver que la ruta `/users` usa la funcion `getUsers` que está definido en
```

`controller/users.js`. El concepto de controller nos sirve para separar más
la logica de la definición de rutas con la implementacion que va a hablar con
el base de datos. Hay que implementar `getUsers` para obtener la lista de

> users del colleción en tu base de datos.

Reviewer: ```suggestion
users de la colección en tu base de datos.
```

de datos. Algunas bases de datos ofrecen herramientas (como
[`mongodb-memory-server`](https://github.com/nodkz/mongodb-memory-server)) que
nos permiten usar una base de datos en memoria y así evitar hacer _mocks_ per
se, pero por lo general querremos considerar cómo abstraer la interacción
con la base de datos para facilitar _mocks_ que nos permitan concentrarnos en
la lógica de las rutas.

## 7. Familiarizarte con las pruebas de integración (e2e)

El _boilerplate_ de este proyecto ya incluye pruebas `e2e` (end-to-end) o de
_integración_, que se encargan de probar nuestra aplicación en conjunto,
através de la interfaz HTTP. A diferencia de las pruebas unitarias, en vez
de importar o requerir un módulo y probar una función de forma aislada, lo que
vamos a hacer es arrancar toda la aplicación, y probarla tal como se usaría en
el mundo real, para ello la aplicación de prueba necesitará una base de datos y
escuchar en un puerto de red.

Para correr pruebas e2e sobre instancia local podemos usar:

```sh
npm run test:e2e
```

Esto levanta la aplicación con `npm start` y corre los tests contra la URL de
esta instancia (por defecto `http://127.0.0.1:8080`). Esto asume que la base de

> datos está disponible.

Reviewer: @unjust de esta frase entiendo q tienes q levantar mongo por fuera, pero del archivo de global setup entiendo q mongo se levanta internamente en memoria a traves de `@shelf/jest-mongodb/lib/setup`

> Los pruebas e2e depende en codigo de (`globalSetup.js`)[../e2e/globalSetup.js]

Reviewer: ```suggestion
Los pruebas e2e dependen del código de (`globalSetup.js`)[../e2e/globalSetup.js]
```

> que `jest` sabe correr primero, antes de los tests.

Reviewer: ```suggestion
que `jest` ejecuta en primer lugar, antes de los tests.
```

En este paso de setup, levantamos un mock db (como hemos hablado de
[`mongodb-memory-server`](https://github.com/nodkz/mongodb-memory-server))
y jest se conecta a este mock db.

> Para este configuracion - usamos `mongodb-memory-server` y un preset (`jest-mongodb`)[https://github.com/shelfio/jest-mongodb].

Reviewer: ```suggestion
Para este configuracion usamos `mongodb-memory-server` y un preset (`jest-mongodb`)[https://github.com/shelfio/jest-mongodb].
```

config), podemos proceder a establecer una conexión con la base de datos
usando el cliente elegido.

### Ejemplo de conexión usando MongoDB Node Driver

```js
const { MongoClient } = require('mongodb');
const config = require("./config");

const client = new MongoClient(config.dbUrl);

async function run() {
  try {
    await client.connect();
    const db = client.db('test');
    // ...
  } finally {
    // Close the database connection when finished or an error occurs
    await client.close();
  }
}
run().catch(console.error);
```

Puedes encontrar mas ejemplos en [la documentación de MongoDB](https://www.mongodb.com/docs/drivers/node/current/).

## 4. Definir esquemas

Como parte del proceso de diseño de nuestra base de datos vamos a tener que
especificar los _esquemas_ de nuestros _modelos_ de datos. Con esto nos
referimos a que tenemos que _describir_ de alguna forma la estructura de las colecciones
que vamos a usar y la _forma_ de los objetos que vayamos a
guardar en esas colecciones.

Puedes usar [la documentación de API](https://app.swaggerhub.com/apis-docs/ssinuco/BurgerQueenAPI/2.0.0)
que describe la estructura de `products`, `orders`, etc. para guiar el diseño.

> ## 5. Implementar primeros TODOs

Author: ```suggestion
## 5. Implementar los primeros TODOs
```

Ahora que ya tenemos un servidor de base de datos, vamos a necesitar elegir un
módulo o librería diseñado para interactuar con nuestra base de datos desde
Node.js. Existen un montón de opciones, pero para este proyecto te recomendamos
usar el [Node.js driver de MongoDB](https://www.mongodb.com/docs/drivers/node/current/)
que es la forma más directa de interactuar con tu instancia de MongoDB.

Hay que instalar el Node.js Driver en este proyecto usando `npm`,

> [revisa los docs por más info](https://www.mongodb.com/docs/drivers/node/current/quick-start/download-and-install/#install-the-node.js-driver).

Reviewer 2: ```suggestion
revisa [la documentación oficial para más información](https://www.mongodb.com/docs/drivers/node/current/quick-start/download-and-install/#install-the-node.js-driver).
```

### TODO 1: Conectar a la base de datos

> En el `connect.js` hay que conectar al base de datos.

Reviewer 2: ```suggestion
En el archivo `connect.js` hay que hacer la conexión con la base de datos.
```

Revisa [el tutorial de Node y express en Mozilla](https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/routes)
que habla de controllers.

> ### TODO 6: Implementar a rota POST `/users`

Reviewer: esto está todo en portugués, no?

Os testes e2e invocam a rota POST /users para adicionar a usuária aos testes.
Portanto, antes de poder executar os testes e2e, essa rota deve funcionar
corretamente.

> 

Reviewer 2: Quizás me marié un poco, pero me parece que esta sección es del archivo en Portugués no?

Author: oops! si
